### PR TITLE
import a folder full of EML files to James

### PR DESCRIPTION
**Pull request by ilkerceng**
[https://issues.apache.org/jira/browse/JAMES-2081](url)
implemented necessary codes, corrected indentations, added tests.

### CODE REVIEW

**Comment by chibenwa** [comment]
We tend to prefix our commit names by tickets:

```
JAMES-2081  	IMPORTEMLFILETOMAILBOX command
```

Moreover, you need to review this history. Does your changesets makes sens alone? I would encourage you to keep maximum 2/3 commits for this.

eg: commit 1 : Add capability to MailboxManagement
commit 2 : Add feature to the CLI

One commit is also OK

File: server/container/cli/src/main/java/org/apache/james/cli/type/CmdType.java [hunk 1]
```diff
@@ -41,6 +41,7 @@
     CREATEMAILBOX("CreateMailbox", "namespace", "user", "name"),
     LISTUSERMAILBOXES("ListUserMailboxes", "user"),
     DELETEMAILBOX("DeleteMailbox", "namespace", "user", "name"),
+    IMPORTEMLFILETOMAILBOX("importemlfiletomailbox", "namespace", "user", "name", "path"),
```
Comment (chibenwa) [inline server/container/cli/src/main/java/org/apache/james/cli/type/CmdType.java]:
```
IMPORTEMLFILETOMAILBOX("importEmlFileToMailbox", "namespace", "user", "name", "path")
```

Snake case makes it easier to read for end user

File: server/container/cli/src/test/java/org/apache/james/cli/ServerCmdTest.java [hunk 2]
```diff
@@ -332,7 +332,24 @@ public void deleteMailboxMappingCommandShouldWork() throws Exception {
         testee.executeCommandLine(commandLine);
         control.verify();
     }
+    
+    @Test
+    public void importEmlFileToMailboxCommandShouldWork() throws Exception {
+        String user = "user@domain";
+        String namespace = "#private";
+        String name = "INBOX.test";
+        String emlpath = ClassLoader.getSystemResource("eml/frnog.eml").getFile();
```
Comment (chibenwa) [inline server/container/cli/src/test/java/org/apache/james/cli/ServerCmdTest.java]: You don't need a real path here. "my/path/to/file.eml" will also work.

File: server/container/cli/src/test/java/org/apache/james/cli/type/CmdTypeTest.java [hunk 3]
```diff
@@ -139,6 +139,11 @@ public void lookupDeleteMailboxShouldReturnEnumValue() {
     }
 
     @Test
+    public void lookupImportEmlFileToMailboxShouldReturnEnumValue() {
+        assertThat(CmdType.lookup("importemlfiletomailbox")).isEqualTo(CmdType.IMPORTEMLFILETOMAILBOX);
```
Comment (chibenwa) [inline server/container/cli/src/test/java/org/apache/james/cli/type/CmdTypeTest.java]: Add one breakline before isEqualTo

File: server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/MailboxManagerManagement.java [hunk 4]
```diff
@@ -142,6 +148,25 @@ public void deleteMailbox(String namespace, String user, String name) {
         }
     }
 
+    @Override
+    public void importEmlFileToMailBox(String namespace, String user, String name, String emlpath) {
+        checkMailboxArguments(namespace, user, name, emlpath);
+        MailboxSession session = null;
```
Comment (chibenwa) [inline server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/MailboxManagerManagement.java]: You need to obtain a session from the mailboxManager: 

mailboxManager.createSystemSession will do the trick

File: server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java [hunk 5]
```diff
@@ -236,6 +245,38 @@ public void deleteMailboxShouldNotDeleteGivenMailboxIfWrongName() throws Excepti
...
+        Mailbox mailbox = new SimpleMailbox(new MailboxPath(MailboxConstants.USER_NAMESPACE, USER, "name"),
+                UID_VALIDITY);
+        inMemoryMapperFactory.createMailboxMapper(session).save(mailbox);
+        String emlpath = ClassLoader.getSystemResource("eml/frnog.eml").getFile();
+        mailboxManagerManagement.importEmlFileToMailBox(MailboxConstants.USER_NAMESPACE, USER, "name", "wrong_path" + emlpath);
+
+        assertThat(inMemoryMapperFactory.getMessageMapper(session).countMessagesInMailbox(mailbox)).isEqualTo(0);
+        Iterator<MailboxMessage> iterator = inMemoryMapperFactory.getMessageMapper(session).findInMailbox(mailbox,
+                MessageRange.all(), null, 1);
+        assertFalse(iterator.hasNext());
```
Comment (chibenwa) [inline server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java]: asserThat(iterator.hasNext()).isFalse();

File: server/container/cli/src/main/java/org/apache/james/cli/ServerCmd.java [hunk 6]
```diff
@@ -266,6 +266,9 @@ private void executeCommand(String[] arguments, CmdType cmdType, PrintStream pri
         case DELETEMAILBOX:
             mailboxProbe.deleteMailbox(arguments[1], arguments[2], arguments[3]);
             break;
+        case IMPORTEMLFILETOMAILBOX:
+            mailboxProbe.importEmlFileToMailBox(arguments[1], arguments[2], arguments[3], arguments[4]);
```
Comment (chibenwa) [inline server/container/cli/src/main/java/org/apache/james/cli/ServerCmd.java]: Sorry it's `mailbox` and not `mailBox`. Could you please fix this case issue?

File: server/container/cli/src/main/java/org/apache/james/cli/probe/impl/JmxMailboxProbe.java [hunk 7]
```diff
@@ -84,6 +84,11 @@ public void deleteMailbox(String namespace, String user, String name) {
     }
 
     @Override
+    public void importEmlFileToMailBox(String namespace, String user, String name, String emlpath) {
+        mailboxManagerManagement.importEmlFileToMailBox(namespace, user, name, emlpath);
```
Comment (chibenwa) [inline server/container/cli/src/main/java/org/apache/james/cli/probe/impl/JmxMailboxProbe.java]: Could you please fix this case issue?

File: server/container/cli/src/test/java/org/apache/james/cli/type/CmdTypeTest.java [hunk 8]
```diff
@@ -139,6 +139,12 @@ public void lookupDeleteMailboxShouldReturnEnumValue() {
     }
 
     @Test
+    public void lookupImportEmlFileToMailboxShouldReturnEnumValue() {
+        assertThat(CmdType.lookup("ImportEmlFileToMailbox"))
```
Comment (chibenwa) [inline server/container/cli/src/test/java/org/apache/james/cli/type/CmdTypeTest.java]: Maybe we can get the name of the command shorter... Maybe `importEml` is enough...

I 'm thinking to poor admins...

File: server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/MailboxManagerManagement.java [hunk 9]
```diff
@@ -142,6 +148,25 @@ public void deleteMailbox(String namespace, String user, String name) {
...
+        try {
+            session = mailboxManager.createSystemSession(user, log);
+            mailboxManager.startProcessingRequest(session);
+            MessageManager messageManager = mailboxManager.getMailbox(new MailboxPath(namespace, user, name), session);
+            InputStream emlFileResourceAsStream = new FileInputStream(emlpath);
```
Comment (chibenwa) [inline server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/MailboxManagerManagement.java]: emlFileAsStream

File: server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/MailboxManagerManagement.java [hunk 10]
```diff
@@ -142,6 +148,25 @@ public void deleteMailbox(String namespace, String user, String name) {
...
+            ComposedMessageId composedMessageId = messageManager.appendMessage(emlFileResourceAsStream, new Date(),
+                    session, false, new Flags());
```
Comment (chibenwa) [inline server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/MailboxManagerManagement.java]: What does false means? We tend to extract values to constant to give them a name.

I believe, if it is isRecent then it should be true.

File: server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java [hunk 11]
```diff
@@ -236,6 +245,38 @@ public void deleteMailboxShouldNotDeleteGivenMailboxIfWrongName() throws Excepti
...
+    @Test
+    public void importEmlFileToMailBoxShouldImportEmlFileToGivenMailbox() throws Exception {
+        Mailbox mailbox = new SimpleMailbox(new MailboxPath(MailboxConstants.USER_NAMESPACE, USER, "name"),
+                UID_VALIDITY);
+        inMemoryMapperFactory.createMailboxMapper(session).save(mailbox);
+        String emlpath = ClassLoader.getSystemResource("eml/frnog.eml").getFile();
+        mailboxManagerManagement.importEmlFileToMailBox(MailboxConstants.USER_NAMESPACE, USER, "name", emlpath);
+
+        assertThat(inMemoryMapperFactory.getMessageMapper(session).countMessagesInMailbox(mailbox)).isEqualTo(1);
+        Iterator<MailboxMessage> iterator = inMemoryMapperFactory.getMessageMapper(session).findInMailbox(mailbox,
```
Comment (chibenwa) [inline server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java]: Here what is null? Your fetch type? Use FetchType.Full instead.

Extract 1 in a constant named limit

File: server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java [hunk 12]
```diff
@@ -236,6 +245,38 @@ public void deleteMailboxShouldNotDeleteGivenMailboxIfWrongName() throws Excepti
...
+    @Test
+    public void importEmlFileToMailBoxShouldNotImportEmlFileWithWrongPathToGivenMailbox() throws Exception {
+        Mailbox mailbox = new SimpleMailbox(new MailboxPath(MailboxConstants.USER_NAMESPACE, USER, "name"),
+                UID_VALIDITY);
+        inMemoryMapperFactory.createMailboxMapper(session).save(mailbox);
+        String emlpath = ClassLoader.getSystemResource("eml/frnog.eml").getFile();
+        mailboxManagerManagement.importEmlFileToMailBox(MailboxConstants.USER_NAMESPACE, USER, "name", "wrong_path" + emlpath);
+
+        assertThat(inMemoryMapperFactory.getMessageMapper(session).countMessagesInMailbox(mailbox)).isEqualTo(0);
+        Iterator<MailboxMessage> iterator = inMemoryMapperFactory.getMessageMapper(session).findInMailbox(mailbox,
```
Comment (chibenwa) [inline server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java]: Idem here

File: server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java (hunk 11 above)
Comment (chibenwa) [inline server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java]: Sorry, named LIMIT

File: server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java (hunk 12 above)
Comment (ilkerceng) [inline server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java]: sorry but what you mean with "Idem here"?

File: server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java (hunk 11 above)
Comment (ilkerceng) [inline server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java]: i have changed it as MessageMapper.FetchType.Full, added LIMIT as a constant for 1

File: server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/MailboxManagerManagement.java (hunk 10 above)
Comment (ilkerceng) [inline server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/MailboxManagerManagement.java]: yep it is isRecent, i have changed isRecent as true.

File: server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/MailboxManagerManagement.java (hunk 9 above)
Comment (ilkerceng) [inline server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/MailboxManagerManagement.java]: i have changed it as emlFileAsStream

File: server/container/cli/src/test/java/org/apache/james/cli/type/CmdTypeTest.java (hunk 8 above)
Comment (ilkerceng) [inline server/container/cli/src/test/java/org/apache/james/cli/type/CmdTypeTest.java]: hmm, what is your last decision for that :) should i change the name "importEmlFileToMailbox" as importEml for all occurences

File: server/container/cli/src/main/java/org/apache/james/cli/probe/impl/JmxMailboxProbe.java (hunk 7 above)
Comment (ilkerceng) [inline server/container/cli/src/main/java/org/apache/james/cli/probe/impl/JmxMailboxProbe.java]: i have fixed it as Maibox.

File: server/container/cli/src/main/java/org/apache/james/cli/ServerCmd.java (hunk 6 above)
Comment (ilkerceng) [inline server/container/cli/src/main/java/org/apache/james/cli/ServerCmd.java]: i have fixed it as Maibox.

File: server/container/cli/src/main/java/org/apache/james/cli/probe/impl/JmxMailboxProbe.java (hunk 7 above)
Comment (chibenwa) [inline server/container/cli/src/main/java/org/apache/james/cli/probe/impl/JmxMailboxProbe.java]: It should be:

```
mailboxManagerManagement.importEmlFileToMailbox(namespace, user, name, emlpath);
```

File: server/container/cli/src/main/java/org/apache/james/cli/type/CmdType.java [hunk 13]
```diff
@@ -41,6 +41,7 @@
     CREATEMAILBOX("CreateMailbox", "namespace", "user", "name"),
     LISTUSERMAILBOXES("ListUserMailboxes", "user"),
     DELETEMAILBOX("DeleteMailbox", "namespace", "user", "name"),
+    IMPORTEMLFILETOMAILBOX("ImportEmlFileToMailbox", "namespace", "user", "name", "path"),
```
Comment (chibenwa) [inline server/container/cli/src/main/java/org/apache/james/cli/type/CmdType.java]: It should be :

```
IMPORTEML("ImportEml", "namespace", "user", "name", "path"),
```

For end user readability

File: server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java (hunk 12 above)
Comment (chibenwa) [inline server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxManagementTest.java]: I mean I have the same comment than above:

```


Here what is null? Your fetch type? Use FetchType.Full instead.

Extract 1 in a constant named limit
```

**Comment by ilkerceng** [comment]
mr. @chibenwa 
you had said before to fix absolute path to relative path in ServerCmdTest.java
i am using ClassLoader.getSystemResource("eml/frnog.eml").getFile() still; in MailboxManagementTest.java
1) i think i should fix all occurences of ClassLoader.getSystemResource("eml/frnog.eml").getFile() to relative path?
2) i have realised that i have forgotten pushing "eml/frnog.eml" resource :(

File: server/container/guice/mailbox/src/main/java/org/apache/james/modules/MailboxProbeImpl.java [hunk 14]
```diff
@@ -178,4 +178,10 @@ public void reIndexAll() throws Exception {
         MailboxSession mailboxSession = mailboxManager.createSystemSession(user, LOGGER);
         return subscriptionManager.subscriptions(mailboxSession);
     }
+
+	//TODO: ilker: impelement this
```
Comment (chibenwa) [inline server/container/guice/mailbox/src/main/java/org/apache/james/modules/MailboxProbeImpl.java]: Could you please do it then ?

File: server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/MailboxManagerManagement.java [hunk 15]
```diff
@@ -167,6 +192,13 @@ private void checkMailboxArguments(String namespace, String user, String name) {
         checkString(name, "mailbox name");
     }
 
+	private void checkMailboxArguments(String namespace, String user, String name, String emlpath) {
+		checkString(namespace, "mailbox path namespace");
+		checkString(user, "mailbox path user");
+		checkString(name, "mailbox name");
+		checkString(emlpath, "email file path name");
```
Comment (strokyl) [inline server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/MailboxManagerManagement.java]: I would remove the checkString private method and do:
```
		Preconditions.checkArgument(!Strings.isNullOrEmpty(namespace), "namespace can't be null or empty");
```
using Strings from Guava: https://google.github.io/guava/releases/18.0/api/docs/com/google/common/base/Strings.html

Comment (chibenwa) [inline server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/MailboxManagerManagement.java]: It's done in https://github.com/linagora/james-project/pull/889

**Comment by chibenwa** [comment]
This has just been merged. Could you please close the pull request? This will avoid asking this to the Apache INFRA team...

Thanks again for your contribution.

**Comment by ilkerceng** [comment]
I also thank you so much.

**Pull request closed without merging**